### PR TITLE
47149_email_implementation

### PR DESCRIPTION
**Pull request by manazirahsan**
This is preliminary email implementation for our project.
Later minor changes will be committed while implementing scheduled jobs.

### CODE REVIEW

**Comment by noor-framgia** [comment]
ok

File: app/mailers/application_mailer.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class ApplicationMailer < ActionMailer::Base
+  default from: 'framgia.bangladesh@gmail.com'
```
Comment (shakil-fg) [inline app/mailers/application_mailer.rb]: Double quote

File: app/mailers/tms_mailer.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    @tasks = tasks
+    tasks.each do |task|
+      mail(to: @user.email, subject: t(:email_trainee_report_subject))
+    end
+  end
+
+  def email_about_deadline user, course
+    @user = user
+    @course = course
+    mail(to: 'framgia.bangladesh@gmail.com', subject: t(:email_about_deadline_subject))
```
Comment (shakil-fg) [inline app/mailers/tms_mailer.rb]: Double quote

File: app/views/tms_mailer/email_trainee_added.html.erb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Hi<h4><%= @user.name %></h4>
+You have been added to course <%= @course.name %>.<br/><br/><br/>
+Please login <%= url_for(controller: 'static_pages',
+                                action: 'home',
+                                only_path: false) %> to see.
```
Comment (shakil-fg) [inline app/views/tms_mailer/email_trainee_added.html.erb]: Double quote

File: app/views/tms_mailer/email_trainee_removed.html.erb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+Hi<h4><%= @user.name %></h4>
+You have been removed to course <%= @course.name %>.<br/><br/><br/>
+Please login <%= url_for(controller: 'static_pages',
+                         action: 'home',
+                         only_path: false) %> to see.
```
Comment (shakil-fg) [inline app/views/tms_mailer/email_trainee_removed.html.erb]: Double quote

File: config/application.rb [hunk 5]
```diff
@@ -25,5 +25,15 @@ class Application < Rails::Application
...
+    config.action_mailer.smtp_settings = {
+        address:              'smtp.gmail.com',
+        port:                 587,
+        domain:               'gmail.com',
+        user_name:            'framgia.bangladesh',
+        password:             'mannamanna',
+        authentication:       'plain',
+        enable_starttls_auto: true  }
   end
 end
```
Comment (shakil-fg) [inline config/application.rb]: Double quote

File: config/application.rb [hunk 6]
```diff
@@ -25,5 +25,15 @@ class Application < Rails::Application
 
     # Do not swallow errors in after_commit/after_rollback callbacks.
     config.active_record.raise_in_transactional_callbacks = true
+    config.action_mailer.default_url_options = { host: 'localhost', port: '3000' }
```
Comment (mdmahmudurrahman) [inline config/application.rb]: Please double quote!

File: config/application.rb [hunk 7]
```diff
@@ -25,5 +25,15 @@ class Application < Rails::Application
...
+    config.action_mailer.delivery_method = :smtp
+    config.action_mailer.smtp_settings = {
+        address:              'smtp.gmail.com',
```
Comment (mdmahmudurrahman) [inline config/application.rb]: Double quote

File: app/views/tms_mailer/email_trainee_added.html.erb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+Hi<h4><%= @user.name %></h4>
+You have been added to course <%= @course.name %>.<br/><br/><br/>
+Please login <%= url_for(controller: 'static_pages',
```
Comment (mdmahmudurrahman) [inline app/views/tms_mailer/email_trainee_added.html.erb]: Please double quote.

File: app/mailers/tms_mailer.rb (hunk 2 above)
Comment (mdmahmudurrahman) [inline app/mailers/tms_mailer.rb]: Please double quote.

File: app/mailers/application_mailer.rb (hunk 1 above)
Comment (mdmahmudurrahman) [inline app/mailers/application_mailer.rb]: Double quote please.

File: app/views/tms_mailer/email_trainee_removed.html.erb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+Hi<h4><%= @user.name %></h4>
+You have been removed to course <%= @course.name %>.<br/><br/><br/>
+Please login <%= url_for(controller: 'static_pages',
```
Comment (mdmahmudurrahman) [inline app/views/tms_mailer/email_trainee_removed.html.erb]: Double quote.

File: app/views/tms_mailer/email_trainee_added.html.erb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+Hi<h4><%= @user.name %></h4>
+You have been Added to course <%= @course.name %>.<br/><br/><br/>
+Please login <%= url_for(controller: "static_pages",
+                         action: "home",
+                         only_path: false) %> to see.
```
Comment (MuhammadTamzid) [inline app/views/tms_mailer/email_trainee_added.html.erb]: i18n above

File: app/views/tms_mailer/email_trainee_removed.html.erb [hunk 11]
```diff
@@ -0,0 +1,5 @@
+Hi<h4><%= @user.name %></h4>
+You have been removed to course <%= @course.name %>.<br/><br/><br/>
+Please login <%= url_for(controller: "static_pages",
+                         action: "home",
+                         only_path: false) %> to see.
```
Comment (MuhammadTamzid) [inline app/views/tms_mailer/email_trainee_removed.html.erb]: i18n above

File: config/application.rb [hunk 12]
```diff
@@ -25,5 +25,15 @@ class Application < Rails::Application
 
     # Do not swallow errors in after_commit/after_rollback callbacks.
     config.active_record.raise_in_transactional_callbacks = true
+    config.action_mailer.default_url_options = { host: "localhost", port: "3000" }
```
Comment (MuhammadTamzid) [inline config/application.rb]: Convention
{host: "localhost", port: "3000"}

File: app/views/tms_mailer/email_trainee_added.html.erb (hunk 10 above)
Comment (manazirahsan) [inline app/views/tms_mailer/email_trainee_added.html.erb]: This is our controller action not output-label.

File: app/views/tms_mailer/email_trainee_removed.html.erb (hunk 11 above)
Comment (manazirahsan) [inline app/views/tms_mailer/email_trainee_removed.html.erb]: This is not output-label

File: config/application.rb (hunk 12 above)
Comment (manazirahsan) [inline config/application.rb]: I understood.
So the convention is not to put space after opening-curly-brace and before ending-curly-brace?

File: config/application.rb [hunk 13]
```diff
@@ -25,5 +25,15 @@ class Application < Rails::Application
...
     # Do not swallow errors in after_commit/after_rollback callbacks.
     config.active_record.raise_in_transactional_callbacks = true
+    config.action_mailer.default_url_options = {host: "localhost", port: "3000"}
+    config.action_mailer.delivery_method = :smtp
+    config.action_mailer.smtp_settings = {
+        address:              "smtp.gmail.com",
+        port:                 587,
+        domain:               "gmail.com",
+        user_name:            "framgia.bangladesh",
+        password:             "mannamanna",
```
Comment (MuhammadTamzid) [inline config/application.rb]: user_name and password should be stored in environment variable

File: app/views/tms_mailer/email_trainee_removed.html.erb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+Hi<h4><%= @user.name %></h4>
```
Comment (MuhammadTamzid) [inline app/views/tms_mailer/email_trainee_removed.html.erb]: i18n

File: app/views/tms_mailer/email_trainee_removed.html.erb [hunk 15]
```diff
@@ -0,0 +1,5 @@
...
+You have been removed to course <%= @course.name %>.<br/><br/><br/>
```
Comment (MuhammadTamzid) [inline app/views/tms_mailer/email_trainee_removed.html.erb]: i18n

File: app/views/tms_mailer/email_trainee_removed.html.erb [hunk 16]
```diff
@@ -0,0 +1,5 @@
...
+Please login <%= url_for(controller: "static_pages",
```
Comment (MuhammadTamzid) [inline app/views/tms_mailer/email_trainee_removed.html.erb]: i18n

File: app/views/tms_mailer/email_trainee_removed.html.erb (hunk 11 above)
Comment (MuhammadTamzid) [inline app/views/tms_mailer/email_trainee_removed.html.erb]: i18n

File: app/views/tms_mailer/email_trainee_added.html.erb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+Hi<h4><%= @user.name %></h4>
```
Comment (MuhammadTamzid) [inline app/views/tms_mailer/email_trainee_added.html.erb]: i18n

File: app/views/tms_mailer/email_trainee_added.html.erb [hunk 18]
```diff
@@ -0,0 +1,5 @@
...
+You have been Added to course <%= @course.name %>.<br/><br/><br/>
```
Comment (MuhammadTamzid) [inline app/views/tms_mailer/email_trainee_added.html.erb]: i18n

File: app/views/tms_mailer/email_trainee_added.html.erb [hunk 19]
```diff
@@ -0,0 +1,5 @@
...
+Please login <%= url_for(controller: "static_pages",
```
Comment (MuhammadTamzid) [inline app/views/tms_mailer/email_trainee_added.html.erb]: i18n

File: app/views/tms_mailer/email_trainee_added.html.erb (hunk 10 above)
Comment (MuhammadTamzid) [inline app/views/tms_mailer/email_trainee_added.html.erb]: i18n

File: app/views/tms_mailer/email_trainee_added.html.erb (hunk 19 above)
Comment (manazirahsan) [inline app/views/tms_mailer/email_trainee_added.html.erb]: vai, this is our controller name.

File: app/views/tms_mailer/email_trainee_removed.html.erb (hunk 16 above)
Comment (manazirahsan) [inline app/views/tms_mailer/email_trainee_removed.html.erb]: Controller name.

File: app/mailers/tms_mailer.rb [hunk 20]
```diff
@@ -0,0 +1,27 @@
+class TmsMailer < ApplicationMailer
+  def email_trainee_added user, course
+    @user = user
+    @course = course
+    mail(to: @user.email, subject: t(:email_inclusion_to_course_subject))
```
Comment (dieunb) [inline app/mailers/tms_mailer.rb]: remove `()`

File: app/mailers/tms_mailer.rb [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  def email_trainee_removed user, course
+    @user = user
+    @course = course
+    mail(to: @user.email, subject: t(:email_removal_from_course_subject))
```
Comment (dieunb) [inline app/mailers/tms_mailer.rb]: remove `()`

File: app/mailers/tms_mailer.rb [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  def email_trainee_report supervisor_id, trainee_id, tasks
+    @user = user
+    @tasks = tasks
+    tasks.each do |task|
+      mail(to: @user.email, subject: t(:email_trainee_report_subject))
```
Comment (dieunb) [inline app/mailers/tms_mailer.rb]: remove `()`

File: app/mailers/tms_mailer.rb [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+    end
+  end
+
+  def email_about_deadline user, course
+    @user = user
+    @course = course
+    mail(to: "framgia.bangladesh@gmail.com", subject: t(:email_about_deadline_subject))
```
Comment (dieunb) [inline app/mailers/tms_mailer.rb]: remove `()`

File: app/views/tms_mailer/email_trainee_added.html.erb [hunk 24]
```diff
@@ -0,0 +1,5 @@
+<%= t :greeting %><h4><%= @user.name %></h4>
+<%= t :inclusion_to_course_text %> <%= @course.name %>.<br/><br/><br/>
```
Comment (dieunb) [inline app/views/tms_mailer/email_trainee_added.html.erb]: use i18n with variable, try it

File: config/environment_variables.yml [hunk 25]
```diff
@@ -0,0 +1,6 @@
+development:
```
Comment (dieunb) [inline config/environment_variables.yml]: don't push this file

File: config/initializers/environment_variables.rb [hunk 26]
```diff
@@ -0,0 +1,13 @@
+module EnvironmentVariablesExample
+  class Application < Rails::Application
+    config.before_configuration do
+      env_file = Rails.root.join("config", 'environment_variables.yml').to_s
```
Comment (dieunb) [inline config/initializers/environment_variables.rb]: double quote

File: config/environment_variables.yml (hunk 25 above)
Comment (manazirahsan) [inline config/environment_variables.yml]: I put a line in .gitignore like bellow:
config/environment_variables.yml 
But it did not work.

Now changed it like below:
/config/environment_variables.yml

**Pull request merged**